### PR TITLE
KFSPTS-23544 adding expense v3 allocation processing and expense report validation

### PR DESCRIPTION
**Pull request by jhulslander**
There is a related non prod SQL PR that needs be reviewed along this

### CODE REVIEW

**Review by ajd299 — CHANGES_REQUESTED** [review]
Code looks good, please see my comments

File: src/test/java/edu/cornell/kfs/concur/eventnotification/rest/jsonObjects/ConcurExpenseDTOJsonTest.java [hunk 1]
```diff
@@ -6,6 +6,7 @@
 import java.io.IOException;
 import java.util.Date;
 
+import org.apache.commons.lang3.StringUtils;
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/test/java/edu/cornell/kfs/concur/eventnotification/rest/jsonObjects/ConcurExpenseDTOJsonTest.java:9]: nope, I shall remove

File: src/test/java/edu/cornell/kfs/concur/eventnotification/rest/jsonObjects/ConcurExpenseDTOJsonTest.java [hunk 2]
```diff
@@ -92,5 +104,27 @@ public void testConcurV3ExpenseListingDTO() throws JsonParseException, JsonMappi
...
+        LOG.info("testConcurExpenseAllocationV3ListingDTO, dto: " + dto.toString());
+        assertEquals(3, dto.getItems().size());
+        
+        ConcurExpenseAllocationV3ListItemDTO allOcationItem = dto.getItems().get(2);
+        assertEquals(new KualiDecimal(50), allOcationItem.getPercentage());
+        assertFalse(allOcationItem.isHidden());
+        assertEquals("6760", allOcationItem.getObjectCode());
+        validateAllocationDetail(allOcationItem.getChart(), IT, ITHACA_CAMPUS, IT);
+        validateAllocationDetail(allOcationItem.getAccount(), S565015, MCNAIR_GRANT, S565015);
+        validateAllocationDetail(allOcationItem.getSubObject(), null, MARYLAND, MARYLAND);
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/test/java/edu/cornell/kfs/concur/eventnotification/rest/jsonObjects/ConcurExpenseDTOJsonTest.java:121]: fixed

File: src/test/java/edu/cornell/kfs/concur/eventnotification/rest/jsonObjects/ConcurExpenseDTOJsonTest.java [hunk 3]
```diff
@@ -92,5 +104,27 @@ public void testConcurV3ExpenseListingDTO() throws JsonParseException, JsonMappi
...
+    }
+    
+    private void validateAllocationDetail(ConcurExpenseAllocationV3ListItemDetailDTO dto, String expectedcode, String expectedValue, String expectedCodeOrValue) {
+        assertEquals(expectedcode, dto.getCode());
+        assertEquals(expectedValue, dto.getValue());
+        assertEquals(expectedCodeOrValue, dto.getCodeOrValue());
+    }
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/test/java/edu/cornell/kfs/concur/eventnotification/rest/jsonObjects/ConcurExpenseDTOJsonTest.java:128]: yes

File: src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurExpenseV3ServiceImpl.java [hunk 4]
```diff
@@ -48,37 +63,121 @@ protected void processExpenseListing(String accessToken, ConcurExpenseV3ListingD
...
+        ConcurAccountInfo info = new ConcurAccountInfo(chart, account, subAccount, objectCode, subObject, project, orgRefId);
+        return info;
+    }
+    
+    protected String getValueFromAllocationDetailDTO(ConcurExpenseAllocationV3ListItemDetailDTO dto) {
+        if (ObjectUtils.isNull(dto)) {
+            return StringUtils.EMPTY;
+        } else {
+            return dto.getCodeOrValue();
+        }
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurExpenseV3ServiceImpl.java:167]: ok

File: src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurExpenseV3ServiceImpl.java [hunk 5]
```diff
@@ -48,37 +63,121 @@ protected void processExpenseListing(String accessToken, ConcurExpenseV3ListingD
         for (ConcurExpenseV3ListItemDTO partialExpenseReportFromListing : expenseList.getItems()) {
             ConcurExpenseV3ListItemDTO fullExpenseReport = getConcurExpenseReport(accessToken,
                     partialExpenseReportFromListing.getId(), partialExpenseReportFromListing.getOwnerLoginID());
-            validateConcurExpenseReport(processingResults, fullExpenseReport);
+            
+            List<ConcurExpenseAllocationV3ListItemDTO> allocationItems = getConcurExpenseAllocationV3ListItemsForReport(accessToken, fullExpenseReport.getId());
+            
+            validateExepsneAllocations(accessToken, processingResults, allocationItems, fullExpenseReport.getId());
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurExpenseV3ServiceImpl.java:69]: ok

File: src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurExpenseV3ServiceImpl.java [hunk 6]
```diff
@@ -48,37 +63,121 @@ protected void processExpenseListing(String accessToken, ConcurExpenseV3ListingD
...
+            String logMessageDetail) {
+        return concurEventNotificationV2WebserviceService.buildConcurDTOFromEndpoint(accessToken,
+                allocationEndpoint, ConcurExpenseAllocationV3ListingDTO.class, logMessageDetail);
     }
 
-    protected void validateConcurExpenseReport(List<ConcurEventNotificationProcessingResultsDTO> processingResults,
-            ConcurExpenseV3ListItemDTO fullExpenseReport) {
-        LOG.info("validateConcurExpenseReport, validation step not implemented yet.");
+
+    protected void validateExepsneAllocations(String accessToken, List<ConcurEventNotificationProcessingResultsDTO> processingResults,
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurExpenseV3ServiceImpl.java:115]: ok

File: src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurExpenseV3ServiceImpl.java [hunk 7]
```diff
@@ -48,37 +63,121 @@ protected void processExpenseListing(String accessToken, ConcurExpenseV3ListingD
...
-        validationMessages.add("Validation not implemented yet");
-        processingResults.add(
-                new ConcurEventNotificationProcessingResultsDTO(ConcurEventNoticationVersion2EventType.ExpenseReport,
-                        ConcurEventNotificationVersion2ProcessingResults.processingError, fullExpenseReport.getId(),
-                        validationMessages));
+        ConcurEventNotificationVersion2ProcessingResults reportResults = ConcurEventNotificationVersion2ProcessingResults.validAccounts;
+        try {
+            for (ConcurExpenseAllocationV3ListItemDTO allocationItem : allocationItems) {
+                ConcurAccountInfo info = buildConcurAccountInfo(allocationItem);
+                LOG.info("validateExepsneAllocations, for report " + reportId + " account info: " + info.toString());
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurExpenseV3ServiceImpl.java:123]: ok

File: src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurExpenseV3ServiceImpl.java [hunk 8]
```diff
@@ -48,37 +63,121 @@ protected void processExpenseListing(String accessToken, ConcurExpenseV3ListingD
...
+                ValidationResult results = concurAccountValidationService.validateConcurAccountInfo(info);
+                reportValid &= results.isValid();
+                validationMessages.addAll(results.getMessages());
+            }
+            if (!reportValid) {
+                reportResults = ConcurEventNotificationVersion2ProcessingResults.invalidAccounts;
+            }
+        } catch (Exception e) {
+            reportValid = false;
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurExpenseV3ServiceImpl.java:132]: I added reportValid as a parameter to updateStatusInConcur

File: src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurExpenseV3ServiceImpl.java [hunk 9]
```diff
@@ -48,37 +63,121 @@ protected void processExpenseListing(String accessToken, ConcurExpenseV3ListingD
...
+                    configurationService.getPropertyValueAsString(ConcurKeyConstants.MESSAGE_CONCUR_EXPENSEV3_EXPENSE_ALLOCATION_LISTING_NEXT_PAGE), reportId);
+            allocationList = getConcurExpenseAllocationV3ListingDTO(accessToken, reportId, allocationList.getNextPage(), 
+                    nextPageLogMessageDetail);
+            allocationItems.addAll(allocationList.getItems());
         }
+        
+        return allocationItems;
+    }
+    
+    protected ConcurExpenseAllocationV3ListingDTO getConcurExpenseAllocationV3ListingDTO(String accessToken, String reportId, String allocationEndpoint,
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurExpenseV3ServiceImpl.java:108]: removed

**Comment by jhulslander** [comment]
I have made changes, this is ready for review again

**Review by ajd299 — APPROVED** [review]
Looks Good, Merging

**Pull request merged**